scripts/zephyr_module.py

```diff
@@ -282,21 +287,41 @@ def git_revision(path):
                                       stderr=None,
                                       cwd=path).wait()
                 if rc:
-                    return revision + '-dirty'
-                return revision
-        return None
+                    return revision + '-dirty', True
+                return revision, False
+        return None, False
 
-    meta_project = {'path': zephyr_base,
-                    'revision': git_revision(zephyr_base)}
-    meta['zephyr'] = meta_project
+    zephyr_revision, zephyr_dirty = git_revision(zephyr_base)
+    zephyr_project = {'path': zephyr_base,
+                      'revision': zephyr_revision}
+    meta['zephyr'] = zephyr_project
+    workspace_dirty |= zephyr_dirty
 
     if west_projects is not None:
+        from west.manifest import MANIFEST_REV_BRANCH
+        projects = west_projects['projects']
         meta_projects = []
-        for project in west_projects['projects']:
-            project_path = PurePath(project).as_posix()
+
+        # Special treatment of first project as it is the manifest project and
+        # thus doesn't have a MANIFEST_REV_BRANCH.
+        manifest_path = PurePath(projects[0].posixpath).as_posix()
+        manifest_revision, manifest_dirty = git_revision(manifest_path)
+        workspace_dirty |= manifest_dirty
+        manifest_project = {'path': manifest_path,
+                            'revision': manifest_revision}
+        meta_projects.append(manifest_project)
+
+        for project in projects[1:]:
+            project_path = PurePath(project.posixpath).as_posix()
+            revision, dirty = git_revision(project_path)
+            workspace_dirty |= dirty
+            if project.sha(MANIFEST_REV_BRANCH) != revision:
+                revision += '-off'
+                workspace_off = True
             meta_project = {'path': project_path,
-                            'revision': git_revision(project_path)}
+                            'revision': revision}
```

Comment:
The difference between the first iteration and the rest is not significant enought to merit the code duplication IMHO.

Let's iterate over an index of projects instead and add a if i == 0 for the special case.

Reply:
it's not only about code duplication, if you notice the project `0` is the manifest repo, and therefore results are stored in a dedicated variables. Those dedicated variables are used further down here:
https://github.com/zephyrproject-rtos/zephyr/blob/aa472a7476157b7ab830744058be62dbd2078fc7/scripts/zephyr_module.py#L351
https://github.com/zephyrproject-rtos/zephyr/blob/aa472a7476157b7ab830744058be62dbd2078fc7/scripts/zephyr_module.py#L352
https://github.com/zephyrproject-rtos/zephyr/blob/aa472a7476157b7ab830744058be62dbd2078fc7/scripts/zephyr_module.py#L357

Also, the code concerning the `MANIFEST_REV_BRANCH` is not to be called for the project `0` because it's the manifest repo and therefore doesn't have that particular branch, so you would endup with something like:
```
        for i, project in enumerate(projects):
            project_path = PurePath(project.posixpath).as_posix()
            revision, dirty = git_revision(project_path)
            workspace_dirty |= dirty
            if i > 0:
                # The manifest project doesn't have a MANIFEST_REV_BRANCH.
                if project.sha(MANIFEST_REV_BRANCH) != revision:
                    revision += '-off'
                    workspace_off = True

            meta_project = {'path': project_path,
                            'revision': revision}
            meta_projects.append(meta_project)
            if i == 0:
                manifest_revision = revision
                manifest_dirty = dirty
                manifest_project = meta_project
```

That is 15 LOC vs current 16 LOC, but you get an additional `enumerate(projects)` because you need to know when you are processing project `0`.

Most code lines saved are just changed into `manifest_<something> = <something>` instead of the clear `git_revision()` and `PurePath()` calls.

I don't really see the gain in changing this code, but will do so if more people prefers the alternative presented.
@mbolivar-nordic do you have any preference ?

Reply:
"therefore results are stored in a dedicated variables. Those dedicated variables are used further down here:"

I did not notice this, I assumed the comment was correct:

        # Special treatment of first project as it is the manifest project and
        # thus doesn't have a MANIFEST_REV_BRANCH.

Then I suggest that either the comment is reduced to "Special treatment of the first project", or, if you prefer, expanded to explain the full set of reasons that it must be treated specially.

Reply:
That comment is not wrong, but you're right that there is more to it than just this.
Will rephrase the comment.

Reply:
done